synapse/storage/search.py

```diff
@@ -289,7 +324,37 @@ def store_search_entries_txn(self, txn, entries):
                 entry.stream_ordering, entry.origin_server_ts,
             ) for entry in entries)
 
-            txn.executemany(sql, args)
+            # inserts to a GIN index are normally batched up into a pending
+            # list, and then all committed together once the list gets to a
+            # certain size. The trouble with that is that postgres (pre-9.5)
+            # uses work_mem to determine the length of the list, and work_mem
+            # is typically very large.
+            #
+            # We therefore reduce work_mem while we do the insert.
+            #
+            # (postgres 9.5 uses the separate gin_pending_list_limit setting,
+            # so doesn't suffer the same problem, but changing work_mem will
+            # be harmless)
+
+            txn.execute("SET work_mem='256kB'")
+            try:
+                txn.executemany(sql, args)
+            except Exception:
+                # we need to reset work_mem, but doing so may throw a new
+                # exception and we want to preserve the original
```

Comment:
Do we? `work_mem` should be reset if the transaction aborts: https://www.postgresql.org/docs/9.4/static/sql-set.html

Reply:
heh. good point :/